File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/ProceduresFragment.java
package com.hhh.paws.ui.petProfile.menu.procedures;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.PopupMenu;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.hhh.paws.R;
import com.hhh.paws.database.model.SurgicalProcedure;
import com.hhh.paws.database.viewModel.ProcedureViewModel;
import com.hhh.paws.databinding.FragmentProceduresBinding;
import com.hhh.paws.ui.petProfile.menu.ItemClickListener;

import java.util.List;

import dagger.hilt.android.AndroidEntryPoint;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;


@AndroidEntryPoint
public class ProceduresFragment extends Fragment {

    private FragmentProceduresBinding _binding = null;
    private FragmentProceduresBinding getBinding() {
        return _binding;
    }

    private RecyclerView recyclerProcedures;
    private FloatingActionButton addProceduresButton;
    private TextView notElemProcedures,addTextView;
    private ImageView addArrow;
    private ProgressBar progressBarProcedures;

    private LiveData<List<SurgicalProcedure>> listLiveData = new MutableLiveData<>();

    private Disposable disposableGet;
    private ProcedureViewModel viewModelProcedure;
    private ProceduresAdapter adapter;
    private String petNameThis;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        _binding = FragmentProceduresBinding.inflate(inflater, container, false);
        viewModelProcedure = new ViewModelProvider(ProceduresFragment.this).get(ProcedureViewModel.class);
        return getBinding().getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        petNameThis = "Котик";

        notElemProcedures = getBinding().notElemProcedures;
        addTextView = getBinding().addTextView;
        addArrow = getBinding().addArrow;

        progressBarProcedures = getBinding().progressBarProcedures;
        progressBarProcedures.setVisibility(View.VISIBLE);

        recyclerProcedures = getBinding().recyclerProcedures;
        initAdapter();
        adapter.setClickListener(new ItemClickListener() {
            @Override
            public void onItemClickListener(Object object) {
                Bundle bundle = new Bundle();
                bundle.putParcelable("procedure", (SurgicalProcedure) object);
                Navigation.findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                        .navigate(R.id.action_nav_procedures_to_detailProceduresFragment, bundle);
            }

            @Override
            public void onItemLongClickListener(Object object, CardView cardView) {
                showPopUp((SurgicalProcedure) object, cardView);
            }
        });

        addProceduresButton = getBinding().addProceduresButton;
        addProceduresButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Navigation.findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                        .navigate(R.id.action_nav_procedures_to_detailProceduresFragment);
            }
        });

        getNewList();
    }

    private void getNewList() {
        disposableGet = viewModelProcedure.getAllProcedures(petNameThis)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(procedures -> {
                    adapter.setDiffer(procedures);
                    progressBarProcedures.setVisibility(View.INVISIBLE);
                    if (adapter.getItemCount() == 0) {
                        notElemProcedures.setVisibility(View.VISIBLE);
                        addArrow.setVisibility(View.VISIBLE);
                        addTextView.setVisibility(View.VISIBLE);
                    } else {
                        notElemProcedures.setVisibility(View.INVISIBLE);
                        addArrow.setVisibility(View.INVISIBLE);
                        addTextView.setVisibility(View.INVISIBLE);
                    }
                }, error -> {
                    Log.d("Procedure", error.getLocalizedMessage());
                });
    }

    private void initAdapter() {
        adapter = new ProceduresAdapter();
        recyclerProcedures.setAdapter(adapter);
        recyclerProcedures.setLayoutManager(new StaggeredGridLayoutManager(1, LinearLayoutManager.VERTICAL));
    }

    private void showPopUp(SurgicalProcedure currentProcedure, CardView cardView) {
        PopupMenu popupMenu = new PopupMenu(this.getContext(), cardView);
        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                if (item.getItemId() == R.id.deleteMenu) {
                    AlertDialog.Builder alertDialog = new AlertDialog.Builder(requireContext());
                    alertDialog.setIcon(R.mipmap.logo_paws);
                    alertDialog.setTitle("deleteQuestion");
                    alertDialog.setPositiveButton("yes", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            viewModelProcedure.deleteProcedure(petNameThis, currentProcedure);
                            getNewList();
                        }
                    });
                    alertDialog.setNegativeButton("no", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i) {
                            dialogInterface.dismiss();
                        }
                    });
                    alertDialog.show();
                    return true;
                }
                return false;
            }
        });
        popupMenu.inflate(R.menu.long_click_menu);
        popupMenu.show();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        disposableGet.dispose();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        _binding = null;
    }
}

File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/ProceduresFragment.kt
package com.hhh.paws.ui.petProfile.menu.procedures

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.PopupMenu
import android.widget.ProgressBar
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.hhh.paws.R
import com.hhh.paws.database.model.SurgicalProcedure
import com.hhh.paws.database.viewModel.ProcedureViewModel
import com.hhh.paws.databinding.FragmentProceduresBinding
import com.hhh.paws.ui.petProfile.menu.ItemClickListener
import com.hhh.paws.util.PetName.name
import com.hhh.paws.util.UiState
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class ProceduresFragment : Fragment() {

    private var binding: FragmentProceduresBinding? = null
    private var recyclerProcedures: RecyclerView? = null
    private var addProceduresButton: FloatingActionButton? = null
    private var notElemProcedures: TextView? = null
    private var addTextView: TextView? = null
    private var addArrow: ImageView? = null
    private var progressBarProcedures: ProgressBar? = null
    private val viewModelProcedure by viewModels<ProcedureViewModel>()
    private var procedureAdapter: ProceduresAdapter? = null
    private var petNameThis: String? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentProceduresBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        petNameThis = name
        notElemProcedures = binding!!.notElemProcedures
        addTextView = binding!!.addTextView
        addArrow = binding!!.addArrow
        progressBarProcedures = binding!!.progressBarProcedures

        progressBarProcedures!!.visibility = View.VISIBLE
        recyclerProcedures = binding!!.recyclerProcedures

        initAdapter()
        procedureAdapter!!.setClickListener(object : ItemClickListener {
            override fun onItemClickListener(it: Any) {
                val bundle = Bundle()
                bundle.putParcelable("procedure", it as SurgicalProcedure)
                findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                    .navigate(R.id.action_nav_procedures_to_detailProceduresFragment, bundle)
            }

            @RequiresApi(api = Build.VERSION_CODES.Q)
            override fun onItemLongClickListener(it: Any, cardView: CardView) {
                showPopUp(it as SurgicalProcedure, cardView)
            }
        })

        addProceduresButton = binding!!.addProceduresButton
        addProceduresButton!!.setOnClickListener {
            findNavController(requireActivity(), R.id.nav_host_fragment_content_vet_passport)
                .navigate(R.id.action_nav_procedures_to_detailProceduresFragment)
        }

        viewModelProcedure.getAllProcedures(petNameThis!!)
        viewModelProcedure.allProcedures.observe(viewLifecycleOwner) {
            when(it) {
                is UiState.Loading -> {
                    progressBarProcedures?.visibility = View.VISIBLE
                }
                is UiState.Success -> {
                    progressBarProcedures?.visibility = View.INVISIBLE
                    procedureAdapter?.setDiffer(it.data)
                    if (procedureAdapter?.itemCount == 0) {
                        notElemProcedures?.visibility = View.VISIBLE
                        addArrow?.visibility = View.VISIBLE
                        addTextView?.visibility = View.VISIBLE
                    } else {
                        notElemProcedures?.visibility = View.INVISIBLE
                        addArrow?.visibility = View.INVISIBLE
                        addTextView?.visibility = View.INVISIBLE
                    }
                }
                is UiState.Failure -> {
                    progressBarProcedures?.visibility = View.INVISIBLE
                    Log.d("UI State", it.error.toString())
                }
            }
        }

        viewModelProcedure.removeProcedure.observe(viewLifecycleOwner) {
            when(it) {
                is UiState.Loading -> {
                    progressBarProcedures?.visibility = View.VISIBLE
                }
                is UiState.Success -> {
                    progressBarProcedures?.visibility = View.INVISIBLE
                    viewModelProcedure.getAllProcedures(petNameThis!!)
                }
                is UiState.Failure -> {
                    progressBarProcedures?.visibility = View.INVISIBLE
                    Log.d("UI State", it.error.toString())
                }
                else -> {
                    progressBarProcedures?.visibility = View.INVISIBLE
                }
            }
        }

    }

    private fun initAdapter() {
        procedureAdapter = ProceduresAdapter()
        recyclerProcedures?.apply {
            adapter = procedureAdapter
            layoutManager = StaggeredGridLayoutManager(1, LinearLayoutManager.VERTICAL)
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    private fun showPopUp(currentProcedure: SurgicalProcedure, cardView: CardView) {
        val popupMenu = PopupMenu(this.context, cardView)
        popupMenu.setOnMenuItemClickListener(PopupMenu.OnMenuItemClickListener { item ->
            if (item.itemId == R.id.deleteMenu) {
                val alertDialog = AlertDialog.Builder(requireContext())
                alertDialog.setIcon(R.mipmap.logo_paws)
                alertDialog.setTitle(R.string.delete_note)
                alertDialog.setPositiveButton(
                    R.string.yes
                ) { _, _ ->
                    viewModelProcedure.deleteProcedure(petNameThis!!, currentProcedure)
                    progressBarProcedures!!.visibility = View.VISIBLE
                }
                alertDialog.setNegativeButton(
                    R.string.no
                ) { dialogInterface, i -> dialogInterface.dismiss() }
                alertDialog.show()
            }
            false
        })
        popupMenu.inflate(R.menu.long_click_menu)
        popupMenu.setForceShowIcon(true)
        popupMenu.show()
    }

    override fun onDestroy() {
        super.onDestroy()
        binding = null
    }
}